aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,2666 @@
+
+        let inactive_shares = table::new<ObservedLockupCycle, pool_u64::Pool>();
+        table::add(&mut inactive_shares, olc_with_index(0), pool_u64::create());
+
+        move_to(&stake_pool_signer, DelegationPool {
+            active_shares: pool_u64::create(),
+            observed_lockup_cycle: olc_with_index(0),
+            inactive_shares,
+            pending_withdrawals: table::new<address, ObservedLockupCycle>(),
+            stake_pool_signer_cap,
+            total_coins_inactive: 0,
+            operator_commission_percentage,
+            add_stake_events: account::new_event_handle<AddStakeEvent>(&stake_pool_signer),
+            reactivate_stake_events: account::new_event_handle<ReactivateStakeEvent>(&stake_pool_signer),
+            unlock_stake_events: account::new_event_handle<UnlockStakeEvent>(&stake_pool_signer),
+            withdraw_stake_events: account::new_event_handle<WithdrawStakeEvent>(&stake_pool_signer),
+            distribute_commission_events: account::new_event_handle<DistributeCommissionEvent>(&stake_pool_signer),
+        });
+
+        // save delegation pool ownership and resource account address (inner stake pool address) on `owner`
+        move_to(owner, DelegationPoolOwnership { pool_address });
+    }
+
+    fun assert_owner_cap_exists(owner: address) {
+        assert!(owner_cap_exists(owner), error::not_found(EOWNER_CAP_NOT_FOUND));
+    }
+
+    fun assert_delegation_pool_exists(pool_address: address) {
+        assert!(delegation_pool_exists(pool_address), error::invalid_argument(EDELEGATION_POOL_DOES_NOT_EXIST));
+    }
+
+    fun coins_to_redeem_to_ensure_min_stake(
+        src_shares_pool: &pool_u64::Pool,
+        shareholder: address,
+        amount: u64,
+    ): u64 {
+        // find how many coins would be redeemed if supplying `amount`
+        let redeemed_coins = pool_u64::shares_to_amount(
+            src_shares_pool,
+            amount_to_shares_to_redeem(src_shares_pool, shareholder, amount)
+        );
+        // if balance drops under threshold then redeem it entirely
+        let src_balance = pool_u64::balance(src_shares_pool, shareholder);
+        if (src_balance - redeemed_coins < MIN_COINS_ON_SHARES_POOL) {
+            amount = src_balance;
+        };
+        amount
+    }
+
+    fun coins_to_transfer_to_ensure_min_stake(
+        src_shares_pool: &pool_u64::Pool,
+        dst_shares_pool: &pool_u64::Pool,
+        shareholder: address,
+        amount: u64,
+    ): u64 {
+        // find how many coins would be redeemed from source if supplying `amount`
+        let redeemed_coins = pool_u64::shares_to_amount(
+            src_shares_pool,
+            amount_to_shares_to_redeem(src_shares_pool, shareholder, amount)
+        );
+        // if balance on destination would be less than threshold then redeem difference to threshold
+        let dst_balance = pool_u64::balance(dst_shares_pool, shareholder);
```

Comment:
Wouldn't it be impossible to have a non-zero balance on destination pool that's below MIN_COINS_ON_SHARES_POOL?